### ISSUE TITLE
Should NSVerify warn if big-M value is too small?

### ISSUE THREAD

**Issue by castrong** [sisl/NeuralVerification.jl#118]
In issue #107  @AvrahamRaviv brought up an example where m = 100 led to a false hold result with NSVerify. With a larger m value the solver worked. Problematic networks are given [here](https://drive.google.com/drive/folders/1uL1ONmoBy5zdgQa_HlSJFc1rvxAZBrhE) and code to replicate the issue is shown below:

```
using NeuralVerification, LazySets, Test, LinearAlgebra, GLPKMathProgInterface
import NeuralVerification: ReLU, Id

mnist1_file = "$(@__DIR__)/networks/mnist_network_new7.nnet"
mnist1 = read_nnet(mnist1_file, last_layer_activation = Id())

input_center = [0.0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,121,254,136,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,13,230,253,248,99,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,4,118,253,253,225,42,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,61,253,253,253,74,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,32,206,253,253,186,9,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,211,253,253,239,69,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,254,253,253,133,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,142,255,253,186,8,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,149,229,254,207,21,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,54,229,253,254,105,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,152,254,254,213,26,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,112,251,253,253,26,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,29,212,253,250,149,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,36,214,253,253,137,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,75,253,253,253,59,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,93,253,253,189,17,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,224,253,253,84,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,43,235,253,126,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,99,248,253,119,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,225,235,49,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]

input_center_label = argmax(NeuralVerification.compute_output(mnist1, input_center))-1

epsilon = 40
upper = input_center .+ epsilon
lower = input_center .- epsilon
clamp!(upper, 0.0, 255.0)
clamp!(lower, 0.0, 255.0)

X = Hyperrectangle(low=lower, high=upper)


A = [0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -1.0, 0.0]'
b = [0.0]

Yc = HPolytope(A, b)
Y = complement(Yc)

problem_mnist = Problem(mnist1, X, Y)

solver_one = NSVerify(m=100)
solver_two = NSVerify(m=10000)
result_one = solve(solver_one, problem_mnist)
result_two = solve(solver_two, problem_mnist)
println("Result one: ", result_one)
println("Result two: ", result_two)

```

Should there be some warning in cases where it's possible if the big-M value chosen is too low? Should we set it by default based on the problem?

**Comment by tomerarnon** [comment]
We can arbitrarily make big M bigger following this, or indeed detect if a given value of M is too low (by comparing it to the variable values after optimization I suppose). I think we can also set M automatically by first computing the layer-wise bounds. Something like the largest radius value in the bounds + epsilon should be "sufficiently large". I think M just needs to be larger than any other variable in the problem in our case. @changliuliu does this make sense? I think computing the bounds shouldn't be expensive enough to make much difference in NSVerify if so.

**Comment by changliuliu** [comment]
Makes sense. The original version of NSVerify didn't provide an approach to determine M. But we can set M to be bigger than the absolute value of any upper or lower bounds on the nodes.

**Comment by castrong** [comment]
Should we still have M be a parameter that can be set or should it only be automatically computed? If it's still user-settable maybe we can throw a warning if it's below the value computed from the bounds?

**Comment by tomerarnon** [comment]
Looks like using the bounds doesn't work. Empirically, on the `small_nnet` from the test suite, the highest absolute bounds from `get_bounds` is 12.1 and does not work (gives false holds), while anything above 26 works. Using pre-activation bounds (manually, by making a new network with only Id layers) gives a value closer to 70 as the max bound, which works fine. However, I think my thinking on this wasn't right from the get-go, since this value is so far from the empirical "minimum viable value" that it's unlikely that this is the right method.

**Comment by AvrahamRaviv** [comment]
@tomerarnon what about mine examples? Did you check the bounds with get_bounds and try differents M?
It can give you more info about the ratio between max bounds and minimum M that works fine.

**Comment by tomerarnon** [comment]
Good idea. I’ll come back to this later and check that too

**Comment by AvrahamRaviv** [comment]
I have more Nets, that return a counterexample for the same problem (meaning the default M =1000 works fine for them). Ill send you them tomorrow, and than we can also compare them to first 2 that return holds.

**Comment by tomerarnon** [comment]
Hmm, not seeing much of a pattern. On `mnist_network_new7.nnet`, the minimum M that works is roughly 1250. The pre/post-activation bounds give values of 72k and 17k respectively; now _both_ far higher than seemingly necessary.

[This stack exchange](https://or.stackexchange.com/questions/231/when-to-use-indicator-constraints-versus-big-m-approaches-in-solving-mixed-int) also suggests that M only needs to be larger than the upper bound on the set of variables. In that case though, I can't explain why exactly that doesn't work on the previous problem.

**Comment by tomerarnon** [comment]
Using pre-activation bounds should be safe, although it is not a very tight lower bound. That's what we decided to go with, along with a one-time warning for user-set `m` values

**Comment by tomerarnon** [comment]
Closed by #120
